### ISSUE TITLE
wc_empty_cart is not always defined when calling forget_session()

### ISSUE THREAD

**Issue by therealgilles** [woocommerce/woocommerce#25388]
##  Prerequisites
None.

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [ ] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [ ] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue
Trigger a do_action('wp_logout') on the wp-login.php page (or a page where frontend_includes() has not been called) when no user is logged in. An example may be a bot trying to log in as 'admin'.

## Expected/actual behavior
I would expect nothing special to happen. Instead I see the following PHP error:

```
Error: Call to undefined function wc_empty_cart() in .../public_html/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php:279
Stack trace:
#0 .../public_html/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php(270): WC_Session_Handler->forget_session()
#1 .../public_html/wp-includes/class-wp-hook.php(288): WC_Session_Handler->destroy_session('')
#2 .../public_html/wp-includes/class-wp-hook.php(312): WP_Hook->apply_filters(NULL, Array)
#3 .../public_html/wp-includes/plugin.php(478): WP_Hook->do_action(Array)
#4 .../public_html/wp-includes/pluggable.php(581): do_action('wp_logout')
#5 .../public_html/wp-content/plugins/better-wp-security/core/lockout.php(801): wp_logout()
#6 ... | file:.../public_html/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php | ln:279
```

My guess is that the frontend functions (including the cart functions) are not loaded when the wp_logout action occurs. A simple fix for this would be:

```
< wc_empty_cart();
> if (function_exists('wc_empty_cart')) wc_empty_cart();
```

Thoughts?

**Comment by claudiosanches** [comment]
> Trigger a do_action('wp_logout') on the wp-login.php page (or a page where frontend_includes() has not been called) when no user is logged in. An example may be a bot trying to log in as 'admin'.

I had no luck reproducing it, can you give more details? Or a more detailed steps?

Also note that by default WooCommerce core only includes `class-wc-session-handler.php` inside `WooCommerce->frontend_includes()`, so I think that maybe a 3rd party plugin or theme is involved trying to include that class in other places that doesn't suppose to.

**Comment by therealgilles** [comment]
@claudiosan: I agree with you. I do not see how it could happen considering the code, i.e. the cart functions are loaded before the session class. I am going to close the issue. I will see if it happens again.

**Comment by jmslbam** [comment]
Just encountered this fatal error when inserting a Woocommerce Highlighted Product Block

```
[15-Dec-2022 15:50:55 UTC] PHP Fatal error:  Uncaught Error: Call to undefined function wc_empty_cart() in /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php:372
Stack trace:
#0 /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php(363): WC_Session_Handler->forget_session()
#1 /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php(99): WC_Session_Handler->destroy_session()
#2 /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php(71): WC_Session_Handler->init_session_cookie()
#3 /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/includes/class-woocommerce.php(862): WC_Session_Handler->init()
#4 /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/src/StoreApi/Utilities/DraftOrderTrait.php(17): WooCommerce->initialize_session()
#5 /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/src/StoreApi/Utilities/QuantityLimits.php(177): Automattic\WooCommerce\StoreApi\Utilities\QuantityLimits->get_draft_order_id()
#6 /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/src/StoreApi/Utilities/QuantityLimits.php(148): Automattic\WooCommerce\StoreApi\Utilities\QuantityLimits->get_remaining_stock(Object(WC_Product_Variable))
#7 /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/src/StoreApi/Utilities/QuantityLimits.php(55): Automattic\WooCommerce\StoreApi\Utilities\QuantityLimits->get_product_quantity_limit(Object(WC_Product_Variable))
#8 /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/src/StoreApi/Schemas/V1/ProductSchema.php(481): Automattic\WooCommerce\StoreApi\Utilities\QuantityLimits->get_add_to_cart_limits(Object(WC_Product_Variable))
#9 /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/src/StoreApi/Routes/V1/Products.php(67): Automattic\WooCommerce\StoreApi\Schemas\V1\ProductSchema->get_item_response(Object(WC_Product_Variable))
#10 /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/packages/woocommerce-blocks/src/StoreApi/Routes/V1/AbstractRoute.php(109): Automattic\WooCommerce\StoreApi\Routes\V1\Products->get_route_response(Object(WP_REST_Request))
#11 /Users/jm/local/iab/app/public/wp-includes/rest-api/class-wp-rest-server.php(1171): Automattic\WooCommerce\StoreApi\Routes\V1\AbstractRoute->get_response(Object(WP_REST_Request))
#12 /Users/jm/local/iab/app/public/wp-includes/rest-api/class-wp-rest-server.php(1018): WP_REST_Server->respond_to_request(Object(WP_REST_Request), '/wc/store/v1/pr...', Array, NULL)
#13 /Users/jm/local/iab/app/public/wp-includes/rest-api/class-wp-rest-server.php(442): WP_REST_Server->dispatch(Object(WP_REST_Request))
#14 /Users/jm/local/iab/app/public/wp-includes/rest-api.php(410): WP_REST_Server->serve_request('/wc/store/v1/pr...')
#15 /Users/jm/local/iab/app/public/wp-includes/class-wp-hook.php(308): rest_api_loaded(Object(WP))
#16 /Users/jm/local/iab/app/public/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters(NULL, Array)
#17 /Users/jm/local/iab/app/public/wp-includes/plugin.php(565): WP_Hook->do_action(Array)
#18 /Users/jm/local/iab/app/public/wp-includes/class-wp.php(399): do_action_ref_array('parse_request', Array)
#19 /Users/jm/local/iab/app/public/wp-includes/class-wp.php(780): WP->parse_request('')
#20 /Users/jm/local/iab/app/public/wp-includes/functions.php(1332): WP->main('')
#21 /Users/jm/local/iab/app/public/wp-blog-header.php(16): wp()
#22 /Users/jm/local/iab/app/public/index.php(17): require('/Users/jm/...')
#23 {main}
  thrown in /Users/jm/local/iab/app/public/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php on line 372
```
<img width="976" alt="Screenshot 2022-12-15 at 20 00 24" src="https://user-images.githubusercontent.com/145887/207945049-b302ca7c-cf51-46cb-bd2d-420c5d502945.png">

**Comment by jmslbam** [comment]
I solved it by adding this as to my theme's `functions.php` as tmp fix..

```
if ( ! function_exists( 'wc_empty_cart' ) ) {
    function wc_empty_cart() {
        if ( ! isset( WC()->cart ) || '' === WC()->cart ) {
            WC()->cart = new WC_Cart();
        }
        WC()->cart->empty_cart( false );
    }
}
```

I'm running WP `6.1` and Woocommerce `7.2.0`

**Comment by jmslbam** [comment]
I found a related issue in the woocommerce-block repo https://github.com/woocommerce/woocommerce-blocks/issues/7537

**Comment by timur987** [comment]
Hey @jmslbam, thanks for the [fix](https://github.com/woocommerce/woocommerce-blocks/issues/7537#issuecomment-1353743113) you've suggested. I enountered the same error but on WooCommerce.com with the request like
```
GET /wp-json/helper/1.0/oauth/authorize
```
(NewRelic [link](https://onenr.io/0PwJYBgM5Q7))

The trace is similar:
```php
Error message:
Uncaught exception 'Error' with message 'Call to undefined function wc_empty_cart()' 
in /var/www/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php:372

 in WC_Session_Handler::forget_session called at /var/www/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php (363)
 in WC_Session_Handler::destroy_session called at /var/www/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php (99)
 in WC_Session_Handler::init_session_cookie called at /var/www/wp-content/plugins/woocommerce/includes/class-wc-session-handler.php (71)
 in WC_Session_Handler::init called at /var/www/wp-content/plugins/woocommerce/includes/class-woocommerce.php (862)
 in WooCommerce::initialize_session called at /var/www/wp-content/plugins/wccom-plugins/site-search/includes/class-wt-algolia-analytics.php (56)
 in WT_Algolia_Analytics::__construct called at /var/www/wp-content/plugins/wccom-plugins/site-search/includes/class-wt-algolia-analytics.php (529)
 in include_once called at /var/www/wp-content/plugins/wccom-plugins/site-search/woothemes-algolia.php (78)
 in WT_Algolia::__construct called at /var/www/wp-content/plugins/wccom-plugins/site-search/woothemes-algolia.php (89)
 in WT_Algolia::get_instance called at /var/www/wp-includes/class-wp-hook.php (308)
 in WP_Hook::apply_filters called at /var/www/wp-includes/class-wp-hook.php (332)
 in WP_Hook::do_action called at /var/www/wp-includes/plugin.php (517)
 in do_action called at /var/www/wp-settings.php (480)
 in require_once called at /var/www/wp-config.php (53)
 in require_once called at /var/www/wp-load.php (50)
 in require_once called at /var/www/wp-blog-header.php (13)
 in require called at /var/www/index.php (17)
```

### Another version of proposed solution

In my case I can't fix it with just using `! is_admin()` as you've suggested but mixed it a bit to look like:
```php
if( ! is_admin() ) {
	include_once WC_ABSPATH . 'includes/wc-cart-functions.php';
	wc_empty_cart();
}
```
in [`class-wc-session-handler.php:372`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-session-handler.php#L372).

### How to reproduce
- Hack the following checks in lines `90` and `98` in [`class-wc-session-handler.php`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-session-handler.php#L90) (otherwise we need a valid but outdated WooCommerce session, it might be a bit complex to make up):
```php
if ( true || $cookie ) {
```
and
```php
if ( true || ! $this->is_session_cookie_valid() ) {
```
- Call `WC()->initialize_session();` from your code
- See an error while calling `wc_empty_cart();` in the line [`372`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-session-handler.php#L372) due to missing the `include_once` call suggested above.

**Comment by jonathansadowski** [comment]
I've added the `status: developer reproduction` issue, since we now have reproduction done on WooCommerce.com.

**Comment by coreymckrill** [comment]
It looks like this was addressed by #36182. Closing this for now, but feel free to reopen if there's still an issue here.
